### PR TITLE
Announce scalafix v0.3

### PR DESCRIPTION
**Pull request by olafurpg**

### CODE REVIEW

File: blog/_posts/2017-02-25-scalafix-v0.3.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+The current strategy in scalafix is to try and resolve as many conflicts as possible on the tree patch level.
+It can be harder to resolve conflicts on the token level since the original intent of the patch is lost.
+Unsolvable conflicts abort the refactoring.
+In the future, we hope to support more advanced conflict resolution strategies.
+
+In a nutshell, a scalafix `Rewrite` is a `scala.meta.Tree => Seq[Patch]` function.
+The tree is backed by the scala.meta semantic API, so the rewrite is able to query for compiler information such as symbols.
+To demonstrate how rewrites are implemented with scalafix v0.3, let's step through an example use-case.
+
+## Example: Xor to Either
```
Comment (BennyHill) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Might be worth mentioning somewhere in the doc (need not be here) that the reverse use case is also possible. i.e a downgrade rewriter.

Users of libraries will tend to want upgrade rewrites that they use once.
Writers of libraries will tend to want to have dynamically called downgrade rewriters to produce mutilple artifacts from the same source

Comment (olafurpg) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Not explicitly, but I sort of mention this possibility here https://github.com/scala/scala-lang/pull/605/files#diff-667c73cf4f57fd817b43faf6b1753bf0R125

File: blog/_posts/2017-02-25-scalafix-v0.3.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+A scalafix `Patch` is a small operation that can produce a diff on a Scala source file.
+A patch can either be a "token patch" or a "tree patch".
+
+Token patches are low-level but give full control over how every detail in a source file is handled, for example formatting and comments.
+Example token patches are `Remove(token)` and `AddLeft(token, toAdd: String)`, which removes or prepends a string to `token`, respectively.
+
+Tree patches are high-level and allow rewrite author to declaratively explain what operation to perform.
+An example tree patch is `AddGlobalImport(importer)`, which adds a new import to the top of a file if it does not exist.
+Observe that `AddGlobalImport` does not worry about token-level details such as whether the user groups imports by prefix (e.g., `import a.{b, c}`).
```
Comment (BennyHill) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: or....something like (this is my language/notes, please change accordingly for your taste, or just ignore it!) :

...as whether the user:
- groups imports by prefix (e.g., `import a.{b, c}`).
- renames imports (e.g., `import a.{b => c, c => b}`).
- groups imports (e.g. `import a.b,  import a.c}`).
- any combinations of these (eg....)

It is  the last point where scalafx shines - trying doing that with regex!

File: blog/_posts/2017-02-25-scalafix-v0.3.md (hunk 1 above)
Comment (BennyHill) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: sure, but that is still downgrading. Up to you... it's your doc, not a major point,

**Comment by BennyHill** [comment]
General comment: In my mind, I think of scalafix as a tool for transforming scala code => dotty. ( I think because of orginal announcements, etc). Of course, we now see we can do more, eg dotty => scala

Whether I'm right or wrong on this point, might be worth adding something to the first paragraph to say what scalafix is :

`I am happy to announce the release of scalafix v0.3, the Scala code rewriting tool and library. This release uses`

**Comment by BennyHill** [comment]
Another general comment, an opinionated one, not well written I admit, so just ignore as you wish. 

I think a small, spoon-feeding example at the start could give some context to the reader. So something along the line of:

- here's some code that does X, but we can code that in different ways depending on the author's preferred style and the version of scala 
- the syntax is different (how we describe what we want to do)
- the semantics are identical - what we want to do is the same

So by mapping the code (a specific syntax) to its semantic representation, we can easily ..... such as as rewriting code from one version to another, even to a version in another language or library

An example could be the package clauses from http://www.artima.com/scalazine/articles/chained_package_clauses_in_scala.html
Note 

> What does this mean? In fact it means exactly the same as

meaning....semantics

> ....change is to allow the new chained package clause syntax

Change.....syntax

> The fix is fortunately straightforward. Simply replace each leading package clause by two or more package

fix...scalafix

**Comment by SethTisue** [comment]
> I think of scalafix as a tool for transforming scala code => dotty

It's important to our plans for Scala 2.13 as well. I suggest that the blog post should reflect this.

**Comment by MasseGuillaume** [comment]
@SethTisue What would be the breaking changes?

**Comment by SethTisue** [comment]
> What would be the breaking changes?

That's still evolving. The major thing that comes to mind is the collections changes.

(But also, we may backport some of the Dotty changes such as the deprecation of procedure syntax we have begun removing deprecated things in 2.13 and might like to offer rewrites in some cases, etc etc etc, it's early yet.)

File: blog/_posts/2017-02-25-scalafix-v0.3.md (hunk 2 above)
Comment (olafurpg) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: I clarified this part. I didn't mention import rename since that is not really handled at the moment.

**Comment by BennyHill** [comment]
:+1:

**Comment by SethTisue** [comment]
thank you

File: blog/_posts/2017-02-25-scalafix-v0.3.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+[@xeno_by][] at Twitter and [myself][@olafurpg] at the Scala Center.
+The objective of the semantic API is to provide operations to query information from the compiler.
+
+The first version of the scala.meta semantic API makes it possible to query for the resolved "symbol" of a name that appears in a Scala source file.
+A name is a reference to some definition, for example `println` or `scala.Predef.println`.
+A symbol is a unique identifier of a single definition.
+For example, `println` from the standard library has the symbol `_root_.scala.Predef.println(Ljava/lang/Object;)V.`.
+The compiler is responsible for resolving names to symbols.
+
+Scalahost is a compiler plugin in the scala.meta project that extracts symbols from the compiler and maps them to scala.meta syntax trees.
```
Comment (heathermiller) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Link to scalahost?

File: blog/_posts/2017-02-25-scalafix-v0.3.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+Scalahost emits the extracted symbols into a "semantic database".
+The semantic database can be persisted to files on disk and loaded for later analysis.
+Semantic databases from different compilation units, potentially produced by different
+versions of the Scala compiler, can be merged.
+This opens possibilities for large-scale code analysis.
+
+The introduction of the scala.meta semantic API is a game changer for scalafix.
+The ability to resolve names to symbols opens possibilities for many scalafix rewrites.
```
Comment (heathermiller) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Maybe add a sentence mentioning that we'll cover a few such examples below?

File: blog/_posts/2017-02-25-scalafix-v0.3.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+## Rewrite: meta.Tree => Seq[Patch]
+
+A scalafix `Patch` is a small operation that can produce a diff on a Scala source file.
+A patch can either be a "token patch" or a "tree patch".
+
+Token patches are low-level but give full control over how every detail in a source file is handled, for example formatting and comments.
+Example token patches are `Remove(token)` and `AddLeft(token, toAdd: String)`, which removes or prepends a string to `token`, respectively.
+
+Tree patches are high-level and allow rewrite author to declaratively explain what operation to perform.
```
Comment (heathermiller) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: > allow rewrite author

should be

>  allow **the** rewrite author

File: blog/_posts/2017-02-25-scalafix-v0.3.md (hunk 3 above)
Comment (olafurpg) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Good idea, I linked to https://github.com/scalameta/sbt-semantic-example

File: blog/_posts/2017-02-25-scalafix-v0.3.md (hunk 4 above)
Comment (olafurpg) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Good idea, I moved the conclusion of the "rewrite" section to the top of the section to make the flow better.

File: blog/_posts/2017-02-25-scalafix-v0.3.md (hunk 5 above)
Comment (olafurpg) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Good catch! Fixed

File: blog/_posts/2017-02-25-scalafix-v0.3.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+```diff
+-final def either: Xor[HCursor, HCursor] = if (succeeded) Xor.right(any) else Xor.left(any)
++final def either: Either[HCursor, HCursor] = if (succeeded) Right(any) else Left(any)
+```
+
+For this particular rewrite, we are able to get away with only using tree patches.
+`Replace` is one tree patch that can be used to replace usage of a reference such as a type, term or a static method.
+
+```scala
+Replace(Symbol("_root_.cats.data.Xor."), q"Either")
```
Comment (olafurpg) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Trailing dot is required, it indicates a `Signature.{Term,Method}`, `Signature.Type` ends with `#` (e.g., "_root_.scala.util.Failure#").
Currently a missing `.` in the suffix results in a runtime error, but we'll probably make `symbol"_root_.foo."` macro interpolator that will work similar to quasiquotes to make this constructor safe.

File: blog/_posts/2017-02-25-scalafix-v0.3.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+I am excited to see what applications the community can build with scala.meta and scalafix.
+Some promising ideas that have floated around include
+
+- parse and run rewrites from `@deprecated` warning messages. This would enable
+  library authors to provide executable migration guides.
+- shim/cross-build libraries with similar APIs. For example,
+  a library could be developed with the scalaz API, and use scalafix to
+  code-generate a version of the library that uses cats instead.
+- build a code search web-interface with "jump to definition" functionality
```
Comment (olafurpg) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Exactly! I added a link to sxr :)

Comment (dwijnand) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: for the 1% of scala developers that know of Scala XRay :)

File: blog/_posts/2017-02-25-scalafix-v0.3.md (hunk 6 above)
Comment (olafurpg) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: I think it's fine to leave out these details in the post.

File: blog/_posts/2017-02-25-scalafix-v0.3.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+[@xeno_by][] at Twitter and [myself][@olafurpg] at the Scala Center.
+The objective of the semantic API is to provide operations to query information from the compiler.
+
+The first version of the scala.meta semantic API makes it possible to query for the resolved "symbol" of a name that appears in a Scala source file.
+A name is a reference to some definition, for example `println` or `scala.Predef.println`.
+A symbol is a unique identifier of a single definition.
+For example, `println` from the standard library has the symbol `_root_.scala.Predef.println(Ljava/lang/Object;)V.`.
+The compiler is responsible for resolving names to symbols.
+
+[Scalahost][] is a compiler plugin in the scala.meta project that extracts symbols from the compiler and maps them to scala.meta syntax trees.
```
Comment (heathermiller) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Ah, broken link here.

Comment (olafurpg) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: It links fine on my local build, see https://github.com/scala/scala-lang/pull/605/files#diff-667c73cf4f57fd817b43faf6b1753bf0R143

Comment (heathermiller) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Hmm, ok. Didn't know this was valid syntax. Merging then!

Comment (olafurpg) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: Cudos to @dwijnand for teaching me [here](https://github.com/olafurpg/scala-lang/pull/1#discussion_r103059014) 🌟

File: blog/_posts/2017-02-25-scalafix-v0.3.md (hunk 7 above)
Comment (dwijnand) [inline blog/_posts/2017-02-25-scalafix-v0.3.md]: @olafurpg FWIW https://github.com/sbt/sxr is the repo "these days" (but not really supported)

**Pull request merged**